solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -346,6 +353,26 @@ public CoreContainer(NodeConfig config, Properties properties, CoresLocator loca
         ExecutorUtil.newMDCAwareCachedThreadPool(
             cfg.getReplayUpdatesThreads(),
             new DefaultSolrThreadFactory("replayUpdatesExecutor")));
+    log.info("Initialized core container as {}", isQueryAggregator ? "query aggregator" : "data node");
+  }
+
+  private void registerCollectionListener() {
+    getZkController().getZkStateReader().registerCloudCollectionsListener(new CloudCollectionsListener() {
+      @Override
+      public void onChange(Set<String> oldCollections, Set<String> newCollections) {
+        // if core is not in newCollections and it exists locally, we delete the core
+        for (SolrCore core : getCores()) {
+          if (!newCollections.contains(core.getName())) {
+            try {
+              log.info("unloading/deleting core {} ", core.getName());
+              unload(core.getName(), true, true, true);
+            } catch (Exception ex) {
+              log.warn("Unable to unlaod core " + core.getName(), ex);
+            }
+          }
+        }
+      }
+    });
```

Comment:
Will this internally update the core zk state for the distributed query to use?

Reply:
Internally, it has watch for "/solr/collections" children and whenever it changes, it generates callback for us. Similarly, it has mechanism for "/solr/collections/xyz/state.json" and "/solr/configs/_FS4".